File: frontend/src/types/portfolio.ts
export interface SubAllocation {
  current_value: number;
  ideal_value: number;
  gap: number;
  target_pct_of_total: number;
}

export interface Category {
  current_value: number;
  ideal_value: number;
  gap: number;
  current_pct: number;
  target_pct: number;
  status: string;
  sub_allocations: Record<string, SubAllocation>;
}

export interface Summary {
  positive_gap_total: number;
  negative_gap_total: number;
  inflow_sufficient: boolean;
}

export interface RebalancePlan {
  ETFs: number;
  MutualFunds: number;
  Stocks: number;
  Metals: number;
  RealEstate: number;
  Crypto: number;
}

export interface PortfolioResponse {
  total_value: number;
  categories: Record<string, Category>;
  summary: Summary;
  next_monthly_sip: number;
  rebalance_plan: RebalancePlan;
}

File: backend/models/Portfolio.py
from typing import Dict

from pydantic import BaseModel


class SubAllocation(BaseModel):
    current_value: float
    ideal_value: float
    gap: float
    target_pct_of_total: float


class Category(BaseModel):
    current_value: float
    ideal_value: float
    gap: float
    current_pct: float
    target_pct: float
    status: str
    sub_allocations: Dict[str, SubAllocation]


class Summary(BaseModel):
    positive_gap_total: float
    negative_gap_total: float
    inflow_sufficient: bool


class RebalancePlan(BaseModel):
    ETFs: float
    MutualFunds: float
    Stocks: float
    Metals: float
    RealEstate: float
    Crypto: float


class PortfolioUpdate(BaseModel):
    total_value: float
    categories: Dict[str, Category]
    summary: Summary
    next_monthly_sip: float
    rebalance_plan: RebalancePlan


class PortfolioResponse(BaseModel):
    total_value: float
    categories: Dict[str, Category]
    summary: Summary
    next_monthly_sip: float
    rebalance_plan: RebalancePlan
